builder/buildManager.go

```diff
@@ -0,0 +1,97 @@
+package builder
+
+import (
+	"assignment-exec/image-builder/configurations"
+	"fmt"
+	"github.com/pkg/errors"
+	"log"
+)
+
+type BuildManager struct {
+	commands     []command
+	undoCommands *stack
+}
+
+type BuildManagerOption func(*BuildManager) error
+
+func NewBuildManager(options ...BuildManagerOption) (*BuildManager, error) {
+	b := &BuildManager{undoCommands: newStack()}
+	for _, opt := range options {
+		if err := opt(b); err != nil {
+			return nil, errors.Wrap(err, "failed to create build manager instance")
+		}
+	}
+	return b, nil
+}
+
+func WithCommands(assgnEnv *assignmentEnv) BuildManagerOption {
+	return func(b *BuildManager) error {
+
+		var commandList []command
+		commandList = append(commandList, &verifyCommand{assgnEnv: assgnEnv})
+		commandList = append(commandList, &writeDockerfileCommand{assgnEnv: assgnEnv})
+		commandList = append(commandList, &buildCommand{assgnEnv: assgnEnv})
+		commandList = append(commandList, &publishCommand{assgnEnv: assgnEnv})
+		b.commands = commandList
+		return nil
+	}
+}
+
+func (builder *BuildManager) ExecuteCommands() error {
+	for _, cmd := range builder.commands {
+
+		builder.undoCommands.push(cmd)
+
+		if err := cmd.execute(); err != nil {
+			if undoErr := builder.UndoCommands(); undoErr != nil {
+				log.Printf("error in undoing operations: %v", undoErr)
+			}
+			return err
```

Comment:
No, that is the error in executing commands. For example, when there is error while building image then it would first undo all the previous commands and then return the build error to main